File: client/python/angzarr_client/upcaster.py
"""Base Upcaster class for event version transformation.

Upcasters transform old event versions to current versions during replay.
They enable schema evolution without breaking existing event stores.

Example usage:
    from angzarr_client import Upcaster, upcasts

    class OrderUpcaster(Upcaster):
        name = "upcaster-order"
        domain = "order"

        @upcasts(OrderCreatedV1, OrderCreated)
        def upcast_created(self, old: OrderCreatedV1) -> OrderCreated:
            return OrderCreated(
                order_id=old.order_id,
                customer_id=old.customer_id,
                total=0,  # New field with default
            )
"""

from __future__ import annotations

import inspect
from abc import ABC
from typing import Callable

from google.protobuf import any_pb2

from .proto.angzarr import types_pb2 as types

__all__ = ["Upcaster", "upcasts"]


def upcasts(from_type: type, to_type: type):
    """Decorator for upcaster transformation methods.

    Registers the method as a handler that transforms from_type to to_type.
    Validates that type hints match the decorator arguments.

    Args:
        from_type: The old event version to transform from.
        to_type: The new event version to transform to.

    Example:
        @upcasts(OrderCreatedV1, OrderCreated)
        def upcast_created(self, old: OrderCreatedV1) -> OrderCreated:
            return OrderCreated(...)

    Raises:
        TypeError: If type hints don't match decorator arguments.
    """

    def decorator(method: Callable) -> Callable:
        # Validate type hints
        import typing

        hints = typing.get_type_hints(method)
        sig = inspect.signature(method)
        params = list(sig.parameters.keys())

        if len(params) < 2:  # self + old event
            raise TypeError(f"{method.__name__}: must have old event parameter")

        old_param = params[1]
        if old_param not in hints:
            raise TypeError(f"{method.__name__}: missing type hint for '{old_param}'")

        if hints[old_param] != from_type:
            raise TypeError(
                f"{method.__name__}: @upcasts({from_type.__name__}, ...) "
                f"doesn't match type hint {hints[old_param].__name__}"
            )

        # Check return type if present
        if "return" in hints and hints["return"] != to_type:
            raise TypeError(
                f"{method.__name__}: @upcasts(..., {to_type.__name__}) "
                f"doesn't match return hint {hints['return'].__name__}"
            )

        method._is_upcaster = True
        method._from_type = from_type
        method._to_type = to_type
        return method

    return decorator


class Upcaster(ABC):
    """Base class for event version transformation.

    Provides:
    - Event dispatch via @upcasts decorated methods
    - Version transformation
    - Descriptor generation for topology discovery

    Subclasses must:
    - Set `name` class attribute (e.g., "upcaster-order")
    - Set `domain` class attribute (domain to handle)
    - Decorate transformation methods with `@upcasts(OldType, NewType)`

    Usage:
        class OrderUpcaster(Upcaster):
            name = "upcaster-order"
            domain = "order"

            @upcasts(OrderCreatedV1, OrderCreated)
            def upcast_created(self, old: OrderCreatedV1) -> OrderCreated:
                return OrderCreated(
                    order_id=old.order_id,
                    customer_id=old.customer_id,
                    total=0,
                )
    """

    name: str
    domain: str
    _dispatch_table: dict[str, tuple[str, type, type]] = {}  # suffix -> (method, from_type, to_type)

    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)

        # Skip validation for abstract intermediate classes
        if inspect.isabstract(cls):
            return

        # Validate required class attributes
        if not getattr(cls, "name", None):
            raise TypeError(f"{cls.__name__} must define 'name' class attribute")
        if not getattr(cls, "domain", None):
            raise TypeError(f"{cls.__name__} must define 'domain' class attribute")

        cls._dispatch_table = cls._build_dispatch_table()

    @classmethod
    def _build_dispatch_table(cls) -> dict[str, tuple[str, type, type]]:
        """Scan for @upcasts methods and build dispatch table."""
        table = {}
        for attr_name in dir(cls):
            attr = getattr(cls, attr_name, None)
            if callable(attr) and getattr(attr, "_is_upcaster", False):
                from_type = attr._from_type
                to_type = attr._to_type
                suffix = from_type.__name__
                if suffix in table:
                    raise TypeError(f"{cls.__name__}: duplicate upcaster for {suffix}")
                table[suffix] = (attr_name, from_type, to_type)
        return table

    def upcast(self, event_any: any_pb2.Any) -> any_pb2.Any:
        """Transform a single event to current version.

        Args:
            event_any: Packed event as google.protobuf.Any

        Returns:
            Transformed event (new Any), or original if no transformation needed.
        """
        type_url = event_any.type_url

        for suffix, (method_name, from_type, to_type) in self._dispatch_table.items():
            if type_url.endswith(suffix):
                # Unpack old event
                old_event = from_type()
                event_any.Unpack(old_event)

                # Transform
                new_event = getattr(self, method_name)(old_event)

                # Pack new event
                new_any = any_pb2.Any()
                new_any.Pack(new_event)
                return new_any

        # No transformation needed
        return event_any

    @classmethod
    def handle(cls, events: list[types.EventPage]) -> list[types.EventPage]:
        """Transform a list of events to current versions.

        Args:
            events: List of EventPages to transform.

        Returns:
            List of EventPages with transformed events.
        """
        upcaster = cls()
        result = []

        for page in events:
            if page.HasField("event"):
                new_event = upcaster.upcast(page.event)
                new_page = types.EventPage(event=new_event, sequence=page.sequence)
                new_page.created_at.CopyFrom(page.created_at)
                result.append(new_page)
            else:
                result.append(page)

        return result

File: client/go/upcaster.go
// Package angzarr provides event version transformation via UpcasterRouter.
package angzarr

import (
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
)

// UpcasterHandler transforms an old event Any to a new event Any.
type UpcasterHandler func(old *anypb.Any) *anypb.Any

// UpcasterRouter transforms old event versions to current versions.
//
// Events matching registered handlers are transformed.
// Events without matching handlers pass through unchanged.
//
// Example:
//
//	router := NewUpcasterRouter("order").
//	    On("OrderCreatedV1", upcastCreatedV1).
//	    On("OrderShippedV1", upcastShippedV1)
//
//	newEvents := router.Upcast(oldEvents)
type UpcasterRouter struct {
	domain   string
	handlers []upcasterEntry
}

type upcasterEntry struct {
	fullName string
	handler  UpcasterHandler
}

// NewUpcasterRouter creates a new upcaster router for a domain.
func NewUpcasterRouter(domain string) *UpcasterRouter {
	return &UpcasterRouter{
		domain:   domain,
		handlers: make([]upcasterEntry, 0),
	}
}

// On registers a handler for an old event type.
//
// The fullName is the fully-qualified proto type name (e.g., "examples.OrderCreatedV1").
// This matches against type_url = "type.googleapis.com/" + fullName.
func (r *UpcasterRouter) On(fullName string, handler UpcasterHandler) *UpcasterRouter {
	r.handlers = append(r.handlers, upcasterEntry{fullName: fullName, handler: handler})
	return r
}

// Upcast transforms a list of events to current versions.
//
// Events matching registered handlers are transformed.
// Events without matching handlers pass through unchanged.
func (r *UpcasterRouter) Upcast(events []*pb.EventPage) []*pb.EventPage {
	result := make([]*pb.EventPage, 0, len(events))

	for _, page := range events {
		event := page.GetEvent()
		if event == nil {
			result = append(result, page)
			continue
		}

		transformed := false
		for _, entry := range r.handlers {
			if event.TypeUrl == TypeURLPrefix+entry.fullName {
				newEvent := entry.handler(event)
				// Clone the page and replace the event
				newPage := proto.Clone(page).(*pb.EventPage)
				newPage.Payload = &pb.EventPage_Event{Event: newEvent}
				result = append(result, newPage)
				transformed = true
				break
			}
		}

		if !transformed {
			result = append(result, page)
		}
	}

	return result
}

// Domain returns the domain this upcaster handles.
func (r *UpcasterRouter) Domain() string {
	return r.domain
}
